drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	return priv->vfio_group_fd;
+}
+
+static int
+virtio_vdpa_device_fd_get(int vid)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	return priv->vfio_dev_fd;
+}
+
+static int
+virtio_vdpa_notify_area_get(int vid, int qid, uint64_t *offset, uint64_t *size)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	struct vfio_region_info reg = { .argsz = sizeof(reg) };
+	int ret;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+
+	ret = virtio_pci_dev_notify_area_get(priv->vpdev, qid, offset, size);
+	if (ret) {
+		DRV_LOG(ERR, "%s fail to get notify area", vdev->device->name);
+		return ret;
+	}
+
+	DRV_LOG(DEBUG, "Vid %d qid:%d offset:0x%"PRIx64"size:0x%"PRIx64, vid, qid,*offset,*size);
+	return 0;
+}
+
+static struct rte_vdpa_dev_ops virtio_vdpa_ops = {
+	.get_queue_num = virtio_vdpa_vqs_max_get,
+	.get_features = virtio_vdpa_features_get,
+	.get_protocol_features = virtio_vdpa_protocol_features_get,
+	.dev_conf = virtio_vdpa_dev_config,
+	.dev_close = virtio_vdpa_dev_close,
+	.set_vring_state = virtio_vdpa_vring_state_set,
+	.set_features = virtio_vdpa_features_set,
+	.migration_done = NULL,
+	.get_vfio_group_fd = virtio_vdpa_group_fd_get,
+	.get_vfio_device_fd = virtio_vdpa_device_fd_get,
+	.get_notify_area = virtio_vdpa_notify_area_get,
+	.get_stats_names = NULL,
+	.get_stats = NULL,
+	.reset_stats = NULL,
+};
+
+static int vdpa_check_handler(__rte_unused const char *key,
+		const char *value, void *ret_val)
+{
+	if (strcmp(value, "1") == 0)
+		*(int *)ret_val = 1;
+	else
+		*(int *)ret_val = 0;
+
+	return 0;
+}
+
+#define VIRTIO_ARG_VDPA 	  "vdpa"
+
+static int
+virtio_pci_devargs_parse(struct rte_devargs *devargs, int *vdpa)
+{
+	struct rte_kvargs *kvlist;
+	int ret = 0;
+
+	if (devargs == NULL)
+		return 0;
+
+	kvlist = rte_kvargs_parse(devargs->args, NULL);
+	if (kvlist == NULL) {
+		DRV_LOG(ERR, "Error when parsing param");
+		return 0;
+	}
+
+	if (rte_kvargs_count(kvlist, VIRTIO_ARG_VDPA) == 1) {
+		/* vdpa mode selected when there's a key-value pair:
+		 * vdpa=1
+		 */
+		ret = rte_kvargs_process(kvlist, VIRTIO_ARG_VDPA,
+				vdpa_check_handler, vdpa);
+		if (ret < 0)
+			DRV_LOG(ERR, "Failed to parse %s", VIRTIO_ARG_VDPA);
+	}
+
+	rte_kvargs_free(kvlist);
+
+	return ret;
+}
+
+static void
+virtio_vdpa_queues_free(struct virtio_vdpa_priv *priv)
+{
+	uint16_t nr_vq = priv->hw_nr_virtqs;
+	struct virtio_vdpa_vring_info *vr;
+	uint16_t i;
+
+	if(priv->vrings) {
```

Comment:
this check is not needed.
if the queues are not allocated, device registration fails.

remove it.

Reply:
it is needed just in case upper code has error, we didn't free null pointer